### PR TITLE
gui: Show encryption status for devices sharing folder (ref #8808)

### PR DESCRIPTION
**Pull request by acolomb**
### Purpose

This re-implements the stalled enhancement from #8808.  Thanks @Craeckie for the idea and first implementation draft!

If a folder is shared to a device with encryption, add a lock icon in front of the device name under "Shared With" in the folder details panel.  Be careful not to add whitespace caused by line wraps in HTML source code, which would defeat the purpose of keeping the icon glued to the name by a non-breaking space.

To make it actually work, change to `overflow-wrap: word-break;` styling, which by the way looks much better than the current wrapping between arbitrary characters, and keeps the superscript numbers attached to the correct word.

More details in the commit messages... (might want to keep when squashing)

### Testing

Tested GUI manually with varying screen widths and especially very long device and folder names.  So far only on Firefox.

### Screenshots

Words only break if they absolutely don't fit the maximum line length:
Screenshot 2024-01-17 at 22-48-14 devtest Syncthing

Same for folder details pane:
Screenshot 2024-01-17 at 22-49-08 devtest Syncthing

The lock icon and superscript number now stay glued to the device name (first / last word):
Screenshot 2024-01-17 at 22-49-33 devtest Syncthing

### CODE REVIEW

**Comment by acolomb** [comment]
See here for a nice write-up of the different wrapping options: https://codersblock.com/blog/deep-dive-into-text-wrapping-and-word-breaking/

@tomasz1986 would appreciate your feedback on the visuals, possibly even cross-browser behavior.  You seem to be well equipped for such testing :)  But do keep in mind that a CSS attribute not being implemented in ancient browsers (IE, you!) is not an argument for me to avoid it in this case.  Rather, broken rendering in broken browsers is their bug and acceptable as long as all modern ones do the right thing.  It's just cosmetics after all and IMHO it's alright for outdated browsers to show their age with modern content.

File: gui/default/assets/css/overrides.css [hunk 1]
```diff
@@ -189,6 +189,11 @@ input[type="checkbox"].extended-attributes-filter-rule-checkbox {
     word-break: break-all;
 }
 
+/* Break up long words in paragraphs only if necessary to prevent text overflow. */
+.overflow-break-word {
+    overflow-wrap: break-word;
```
Comment (tomasz1986) [inline gui/default/assets/css/overrides.css:194]:
```suggestion
    word-wrap: break-word;
```

Can we use `word-wrap` for backwards compatibility? It's the original name of the property and works in every browser possible.

I also thought that maybe it could be possible to replace `word-break: break-all` with `word-wrap: break-word` altogether, but no, it doesn't play nicely with long paths in the Recent Changes modal, pushing the date out of the modal boundaries.

**Comment by tomasz1986** [comment]
> @tomasz1986 would appreciate your feedback on the visuals, possibly even cross-browser behavior. You seem to be well equipped for such testing :) But do keep in mind that a CSS attribute not being implemented in ancient browsers (IE, you!) is not an argument for me to avoid it in this case. Rather, broken rendering in broken browsers is their bug and acceptable as long as all modern ones do the right thing. It's just cosmetics after all and IMHO it's alright for outdated browsers to show their age with modern content.

The visuals definitely look better for me, as currently the list with a lot of folders/devices is basically unreadable!

However, all the `<!-- Avoid stray space...` comments don't look very nice. Is there no prettier alternative for them?

File: gui/default/assets/css/overrides.css (hunk 1 above)
Comment (acolomb) [reply gui/default/assets/css/overrides.css:194]: Are the two really equivalent though?  From the quoted source, I got the impression that `overflow-wrap` allows breaking long words, while otherwise keeping the "regular" word-wrapping intact.

Do you have an overview of browser (in)compatibility?

Comment (tomasz1986) [reply gui/default/assets/css/overrides.css:194]: Mozilla says `overflow-wrap` and the old `word-wrap` are just aliases.

> The property was originally a nonstandard and unprefixed Microsoft extension called word-wrap, and was implemented by most browsers with the same name. It has since been renamed to overflow-wrap, with word-wrap being an alias.

Source: https://developer.mozilla.org/docs/Web/CSS/overflow-wrap

**Comment by acolomb** [comment]
> However, all the `<!-- Avoid stray space...` comments don't look very nice. Is there no prettier alternative for them?

I've searched around a bit and this was the most self-explanatory solutions I could find.  Another option is putting the white-space within the HTML tag and wrapping the closing angle bracket to the next line.  That totally screws up my auto-indentation though and does not even try to give a hint that it was a deliberate choice.

File: gui/default/assets/css/overrides.css (hunk 1 above)
Comment (acolomb) [reply gui/default/assets/css/overrides.css:194]: https://stackoverflow.com/a/61866696/9294540

Comment (acolomb) [reply gui/default/assets/css/overrides.css:194]: So MDN lists no browser that doesn't understand `overflow-wrap`, right?  I'll keep the standardized name then, since that's more future-proof and the Right Thing.  If IE11 wraps them in a funky fashion, nothing is lost and people will get a feeling of how old their software stack is.

**Comment by tomasz1986** [comment]
One problem I've had with `break-word` is that it only considers actual text as words. In my case, I've got hyphens, slashes, etc. in device and folder names, and using `break-word` instead of `break-all` makes them wrap on those hyphens, even though there is enough space to display the whole thing on the same line.

For example, if the folder name is `Work/Documents/Questionnaire-2023`, it likes to break at `-`, especially if there are more folders named like that one after another, which looks weird.

File: gui/default/assets/css/overrides.css (hunk 1 above)
Comment (tomasz1986) [reply gui/default/assets/css/overrides.css:194]: The full support is listed at https://caniuse.com/wordwrap (the MDN is limited in terms of its browser list). Personally, I don't really see a point in using `overflow-wrap` if `word-wrap` is the exact same thing with 100% browser support... Just seems like a property name change for the sake of changing the name.

**Comment by acolomb** [comment]
So what would you expect with your names containing hyphens?  Wrap the whole `Quesionnaires-2023` to the next line?  IMO it's just fine to take the hyphen as a line breaking opportunity.  It should only do that if the line length ends before `-2023,` though (including the trailing comma).

File: gui/default/assets/css/overrides.css (hunk 1 above)
Comment (acolomb) [reply gui/default/assets/css/overrides.css:194]: Well, one is CSS and the other is random technical debt.  When writing new code, we shouldn't use stuff that's only kept for compatibility with past mistakes, *if* there is a proper replacement.

**Comment by tomasz1986** [comment]
> So what would you expect with your names containing hyphens? Wrap the whole `Quesionnaires-2023` to the next line? IMO it's just fine to take the hyphen as a line breaking opportunity. It should only do that if the line length ends before `-2023,` though (including the trailing comma).

The problem is that with a few folders named this way, it wraps them like this:

```
Quesionnaires-
2015, Quesionnaires-
2016, Quesionnaires-
2017, Quesionnaires-
2018, Quesionnaires-
2019, Quesionnaires-
2020, Quesionnaires-
2021, Quesionnaires-
2022, Quesionnaires-
2023

**Comment by acolomb** [comment]
Isn't that expected and correct if a single name does not fit within the line? The alternative would be to start each entry on a new line and almost double the vertical height.

File: gui/default/assets/css/overrides.css (hunk 1 above)
Comment (tomasz1986) [reply gui/default/assets/css/overrides.css:194]: Both are actually part of the official CSS3 specification:

https://www.w3.org/TR/css-text-3/#overflow-wrap-property

> For legacy reasons, UAs must treat [word-wrap](https://www.w3.org/TR/css-text-3/#propdef-word-wrap) as a [legacy name alias](https://www.w3.org/TR/css-cascade-5/#legacy-name-alias) of the [overflow-wrap](https://www.w3.org/TR/css-text-3/#propdef-overflow-wrap) property.

Comment (acolomb) [reply gui/default/assets/css/overrides.css:194]: I stand corrected.  Still, what use case exactly do we need to use the legacy name for?  If the non-legacy variant works everywhere, I'd like to avoid introducing ancient "legacy" stuff in 2024.

Comment (tomasz1986) [reply gui/default/assets/css/overrides.css:194]: To maintain the same level of backwards compatibility? Especially if, as in this specific case, doing so requires literally zero effort. I mean, it does make sense to use `overflow-wrap` if the rest of the code doesn't support old browsers already, but this isn't the case with the Web GUI.

> I'd like to avoid introducing ancient "legacy" stuff in 2024.

It's the whole Web GUI code "legacy" at this point?

Comment (acolomb) [reply gui/default/assets/css/overrides.css:194]: Done.

**Comment by imsodin** [comment]
Are you happy with this as is @tomasz1986 and @acolomb ?

**Review by tomasz1986 — APPROVED** [review]

**Comment by acolomb** [comment]
Yes, I am.

File: gui/default/index.html [hunk 2]
```diff
@@ -585,12 +585,16 @@ <h4 class="panel-title">
...
-                              <span ng-switch-default>{{deviceName(devices[device.deviceID])}}<span ng-if="!$last">,</span></span>
+                            <span ng-if="folder.type !== 'receiveencrypted' && device.encryptionPassword" class="text-nowrap">
+                              <span class="fa fa-lock"></span>&nbsp;<!-- Avoid stray space...
+                         --></span><!-- Avoid stray space...
+                         --><span ng-switch="completion[device.deviceID][folder.id].remoteState"><!-- Avoid stray space...
+                           --><span ng-switch-when="notSharing" data-original-title="{{'The remote device has not accepted sharing this folder.' | translate}}" tooltip>{{deviceName(devices[device.deviceID])}}<sup>1</sup></span><!-- Avoid stray space...
+                           --><span ng-switch-when="paused" data-original-title="{{'The remote device has paused this folder.' | translate}}" tooltip>{{deviceName(devices[device.deviceID])}}<sup>2</sup></span><!-- Avoid stray space...
+                           --><span ng-switch-default>{{deviceName(devices[device.deviceID])}}</span><!-- Avoid stray space...
```
Comment (imsodin) [inline gui/default/index.html:596]: optional:  
This adds a unrelated noise to the diff of the functional change - ideally these space-avoidance-comme\nts (and the tab to space indentation change somewhere above) would be their own PR/commit.

File: gui/default/syncthing/core/syncthingController.js [hunk 3]
```diff
@@ -2662,6 +2662,15 @@ angular.module('syncthing.core')
                          + '&device=' + encodeURIComponent(deviceID));
         };
 
+        $scope.sharesFolderEncrypted = function (folderID, deviceID) {
```
Comment (imsodin) [inline gui/default/syncthing/core/syncthingController.js:2665]: I think the naming scheme of the following functions is easier to understand (e.g. `folderHasUnacceptedDevices`), so something like `folderIsSharedEncrypted`.

**Review by imsodin — COMMENTED** [review]
Thanks for letting me know, I wanted to know it's ready before having a quick look over. Few minor, mostly optional (labeled as such) comments.

Optional suggestion/observation (as in address if you agree or just say no otherwise :) ):  
Until now we used the super-scripts to attach additional information about a remote to a folder. Now there's a combination of a symbol plus superscripts. While symbols are nice, a mixture seems inconsistent/confusing.

**Comment by acolomb** [comment]
I think the mixture does make sense, as sharing a folder encrypted is a permanent configuration choice, while the superscripts indicate temporary conditions that keep it from syncing. So ideally, we don't want to see any superscript numbers, but only icons possibly.

Regarding the PR refactoring, I'll look into it when I get a chance.

File: gui/default/syncthing/core/syncthingController.js (hunk 3 above)
Comment (acolomb) [reply gui/default/syncthing/core/syncthingController.js]: Done.

File: gui/default/index.html (hunk 2 above)
Comment (acolomb) [reply gui/default/index.html:596]: Will look into it and make a separate PR if it can be split off sensibly.

Comment (acolomb) [reply gui/default/index.html:596]: This PR was rebased on top of #9438, implementing the desired separation.  Besides the commit history, the changes are identical.
